File: backend/src/main/java/de/rubeen/bsc/entities/web/RoomEntity.java
package de.rubeen.bsc.entities.web;

import com.fasterxml.jackson.annotation.JsonGetter;

import java.util.Collection;
import java.util.List;

import static java.text.MessageFormat.format;

public class RoomEntity implements Comparable<RoomEntity> {
    private int roomId;
    private String roomName;
    private int roomSize;
    private List<EquipmentEntity> equipments;

    public RoomEntity() {
    }

    public RoomEntity(int roomId, String roomName, int roomSize, List<EquipmentEntity> equipments) {
        this.roomId = roomId;
        this.roomName = roomName;
        this.roomSize = roomSize;
        this.equipments = equipments;
    }

    @JsonGetter
    public int getId() {
        return roomId;
    }

    public void setRoomId(int roomId) {
        this.roomId = roomId;
    }

    @JsonGetter
    public String getName() {
        return roomName;
    }

    public void setRoomName(String roomName) {
        this.roomName = roomName;
    }

    @JsonGetter
    public int getSize() {
        return roomSize;
    }

    public void setRoomSize(int roomSize) {
        this.roomSize = roomSize;
    }

    public List<EquipmentEntity> getEquipments() {
        return equipments;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof RoomEntity) {
            return ((RoomEntity) obj).roomId == this.roomId;
        }
        return false;
    }

    @Override
    public int compareTo(RoomEntity o) {
        return this.roomId - o.roomId;
    }

    public void setEquipments(Collection<EquipmentEntity> equipmentEntities) {
        this.equipments = List.copyOf(equipmentEntities);
    }

    @Override
    public String toString() {
        return format("room: [id: {0}, name: {1}, size: {2}, equipments: {3}]",
                roomId, roomName, roomSize, equipments);
    }

    public static class EquipmentEntity {
        private int equipId;
        private String equipName;

        public EquipmentEntity() {
        }

        public EquipmentEntity(int equipId, String equipName) {
            this.equipId = equipId;
            this.equipName = equipName;
        }

        @JsonGetter
        public int getId() {
            return equipId;
        }

        public void setEquipId(int equipId) {
            this.equipId = equipId;
        }

        @JsonGetter
        public String getName() {
            return equipName;
        }

        public void setEquipName(String equipName) {
            this.equipName = equipName;
        }

        @Override
        public String toString() {
            return format("[id: {0}, name: {1}]", equipId, equipName);
        }
    }
}

File: backend/src/main/java/de/rubeen/bsc/entities/web/RoomEntity.kt
package de.rubeen.bsc.entities.web

import com.fasterxml.jackson.annotation.JsonGetter

import java.text.MessageFormat.format

class RoomEntity : Comparable<RoomEntity> {
    var id: Int = 0
        private set
    var name: String? = null
        private set
    var size: Int = 0
        private set
    var equipments: List<EquipmentEntity>? = null
        private set

    constructor() {}

    constructor(roomId: Int, roomName: String, roomSize: Int, equipments: List<EquipmentEntity>) {
        this.id = roomId
        this.name = roomName
        this.size = roomSize
        this.equipments = equipments
    }

    fun setRoomId(roomId: Int) {
        this.id = roomId
    }

    fun setRoomName(roomName: String) {
        this.name = roomName
    }

    fun setRoomSize(roomSize: Int) {
        this.size = roomSize
    }

    override fun equals(obj: Any?): Boolean {
        return if (obj is RoomEntity) {
            obj.id == this.id
        } else false
    }

    override fun compareTo(o: RoomEntity): Int {
        return this.id - o.id
    }

    fun setEquipments(equipmentEntities: Collection<EquipmentEntity>) {
        this.equipments = equipmentEntities.toList()
    }

    override fun toString(): String {
        return "(id: $id, name: $name, size: $size, equipments: $equipments)"
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + size
        result = 31 * result + (equipments?.hashCode() ?: 0)
        return result
    }

    class EquipmentEntity {
        var id: Int = 0
            private set
        var name: String? = null
            private set

        constructor() {}

        constructor(equipId: Int, equipName: String) {
            this.id = equipId
            this.name = equipName
        }

        fun setEquipId(equipId: Int) {
            this.id = equipId
        }

        fun setEquipName(equipName: String) {
            this.name = equipName
        }

        override fun toString(): String {
            return "(id: $id, name: $name)"
        }
    }
}
